executor/aggregate.go

```diff
@@ -14,27 +14,85 @@
 package executor
 
 import (
+	"sync"
+
 	"github.com/juju/errors"
 	"github.com/pingcap/tidb/expression"
 	"github.com/pingcap/tidb/expression/aggregation"
 	"github.com/pingcap/tidb/mysql"
+	"github.com/pingcap/tidb/sessionctx"
 	"github.com/pingcap/tidb/sessionctx/stmtctx"
 	"github.com/pingcap/tidb/types"
 	"github.com/pingcap/tidb/util/chunk"
 	"github.com/pingcap/tidb/util/codec"
 	"github.com/pingcap/tidb/util/mvmap"
+	"github.com/spaolacci/murmur3"
 	"golang.org/x/net/context"
+	"runtime"
 )
 
 type aggCtxsMapper map[string][]*aggregation.AggEvaluateContext
 
+type aggWorker struct {
+	*sync.WaitGroup
+
+	finishCh       <-chan struct{}
+	aggFuncs       []aggregation.Aggregation
+	groupByItems   []expression.Expression
+	groupValDatums []types.Datum
+	groupKey       []byte
+	groupVals      [][]byte
+	maxChunkSize   int
+}
+
+// HashAggPartialWorker does the following things:
```

Comment:
Could you draw a graph here to describe how the HashAggPartialWorker and HashAggFinalWorker work.